### DISCUSSION TITLE
Efficient saving of editor content in the database

### DISCUSSION THREAD

**Discussion by NiclasDev63** [ueberdosis/tiptap#5677] [General Questions] [answered]
Hi,

I am currently implementing the backend to save the editor content. However, the question arises as to the best way to do this. Currently I send the complete editor content as JSON to my backend after each keystroke (with debounce) and store it in my DB, which is obviously very expensive for large documents. Is there a more efficient way to do this? I know that Notion stores each block individually and so only needs to send the diffs to the backend and only updates individual blocks, but they use a different scheme where each block has a pointer to its parent and its content, which is not the case with Tiptap. Is there a way to update the data in a similarly efficient way without having to change the complete schema of Tiptap? 

I would be grateful for any suggestions.

Thanks

**Comment by nperez0111** [comment, marked as answer]
Hi @NiclasDev63 

There are a couple of different solutions for saving back editor content from a Tiptap editor:
 - Depending on your user experience, you can have the user confirm that they want to save the content to your system (like a save button) so that you incur this storage performance only once
  - If you want to use auto-saving, you can do it with a debounced on change handler that will only save it after some amount of time has passed since the latest change

Converting editor content into JSON is not an expensive process, and for the size of documents that users typically write, the size of documents are typically negligible (measured in KBs), so I would not be so concerned about it until you actually run into the performance issue which I doubt you have.

If you really do need the performance, then you can go ahead and build some sort of diffing system that would be able to take the parts of the document that changed and only send those as updates to the current document. This is obviously a much more complex system to manage so I really would not recommend it unless you see real performance problems over the wire.

**Reply by NiclasDev63** [reply]
First of all, thanks for the quick reply.
Auto-saving should be used in any case and as described, I already use a debounced on change handler, but after each change (and a short waiting time) the whole document, which can easily have 500KB-1MB+, is sent. Since this can quickly happen 100+ times within a short period of time, I thought it wouldn't hurt to look around for a more efficient variant, but then I'll stick with my current variant for now.

**Reply by Dari4sho** [reply]
Hey @NiclasDev63 

I just implemented this for an app and made it work using yjs by using the [Collaboration](https://tiptap.dev/docs/editor/extensions/functionality/collaboration) extension. It is useful for this use case even if you don't want / need collaboration features.

It needs a bit of work to make it run seamlessly, but then you have a mechanism that only sends a few bytes over the wire per update - instead of the whole document.

How it roughly works - instead of (or additionally to) storing the page contents as raw HTML or JSON in your database, you store a binary snapshot of a yjs document (I store it in a `BYTEA` column in Postgres).

On fresh page load, I send the binary data (as a Buffer) to the frontend for a local working copy.  
On the frontend, I merge this Buffer with a fresh ydoc instance (`const ydoc = new Y.Doc()` and `Y.applyUpdate(ydoc, new Uint8Array(serverBuffer))`).

You register an update listener on the ydoc (`ydoc.on('update', (update) => { ... })`) and you send the update to the server as binary data (converted to a Buffer for transmission).  
On the server you do the same. You create a fresh ydoc, get the yjs buffer from the database - merge it into the ydoc - merge the update (both with `Y.applyUpdate`), encode the full state again (with `Y.encodeStateAsUpdate(ydoc)`), and store it back in the database.

This way, you only send around the update of a few bytes and always have the full document on the server / database.

As mentioned, this needs a bit of work and testing around it and adds some complexity.
- For example, to implement debouncing, I had to create an `updateQueue` array that each update (from `ydoc.on('update', ...)`) gets pushed to and ultimately merged in the debounced action (`Y.mergeUpdates(updateQueue)`) before sending it to the server - including cleaning it up after each debounced transmission.
- I recommend registering some `beforeunload` & component unmount mechanism and using the `keepalive` flag for `fetch` operations so you make sure that updates get sent immediately when a user refreshes or changes pages so you reduce the chance of data loss.

It took me about 1 day to implement it including research, trial and error, and testing.  
Bonus point: if you ever want to implement collaboration, this way you'd have a good base to make it work as you're already using yjs.
